### PR TITLE
Candidate PR Review: Bam reader

### PR DESCRIPTION
**Pull request by bvaisvil**
Draft pull request for adding a BAM parser. Still deciding what the best approach is. Current version involves minimal code additions by using current SAM parser and bgzf reader. It currently takes quite a bit longer than the SAM reader (2-3x) primarily due to tag/auxiliary data handling. I will investigate approaches to speeding this up, though dynamic lengths in the tag data may make this difficult within Python

Still need to do:

Add tests
Add documentation

### CODE REVIEW

File: Bio/Align/bam.py [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+def _to_uint(x):
+    """Converts raw bytes in a BAM file to an unsigned integer"""
+    return int.from_bytes(x, byteorder="little", signed=False)
+
+
+class _ReferenceMetadata(NamedTuple):
+    name: str
+    length: int  # not used?
+
+
```
Comment (XinwuYang) [inline Bio/Align/bam.py:22]: #Object for storing alignment info, which is broken down to bytes. May need further clarification on what each field means.

File: Bio/Align/bam.py [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+    type: bytes
+    data: Union[bytes, int, float, array.array]
+
+
+class _BAMToSamReader(bgzf.BgzfReader):
+    """Limited translation reader from BAM to SAM which allows seek(0), read(0), and iteration"""
+
+    _references: Optional[list[_ReferenceMetadata]]
+    _header_iterator: Optional[Iterator[str]]
+
```
Comment (XinwuYang) [inline Bio/Align/bam.py:47]: Would appreciate more insight on the encoding and mappings. Would be nice to include reference links or documents.

File: Bio/Align/bam.py [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+    next_pos: int
+    tlen: int
+
+
+class _TagData(NamedTuple):
+    type: bytes
+    data: Union[bytes, int, float, array.array]
+
+
+class _BAMToSamReader(bgzf.BgzfReader):
```
Comment (XinwuYang) [inline Bio/Align/bam.py:42]: What's the input and output type for seek and read function. Please clarify.

File: Bio/Align/bam.py [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+            data_type = aux_bytes.read(1)
+
+            read_length = cls.read_lengths.get(data_type)
+            if read_length is not None:
+                # Single numeric type, like tag|type|data
+                python_data_type = cls.type_format_mapping[data_type]
+                data_bytes = aux_bytes.read(read_length)
+                data = struct.unpack(python_data_type, data_bytes)[0]
+                aux_data[tag] = _TagData(data_type, data)
+            elif data_type in [b"Z", b"H"]:
```
Comment (XinwuYang) [inline Bio/Align/bam.py:138]: Suggest adding edge cases to test.

File: Bio/Align/bam.py [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+from Bio import bgzf
+from io import StringIO, BytesIO
+from typing import NamedTuple, Optional, Literal, Iterator, Any, Union
+import struct
+import array
+from Bio.Align import sam
+import os
+
+
+def _to_uint(x):
```
Comment (bsanchez-myriad) [inline Bio/Align/bam.py:13]: missing a type hint

Comment (bsanchez-myriad) [reply Bio/Align/bam.py:13]: add in type

File: Bio/Align/bam.py [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+    """Converts raw bytes in a BAM file to an unsigned integer"""
+    return int.from_bytes(x, byteorder="little", signed=False)
+
+
+class _ReferenceMetadata(NamedTuple):
```
Comment (bsanchez-myriad) [inline Bio/Align/bam.py:18]: add documentation

File: Bio/Align/bam.py [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+    tlen: int
+
+
+class _TagData(NamedTuple):
+    type: bytes
+    data: Union[bytes, int, float, array.array]
+
+
+class _BAMToSamReader(bgzf.BgzfReader):
+    """Limited translation reader from BAM to SAM which allows seek(0), read(0), and iteration"""
```
Comment (bsanchez-myriad) [inline Bio/Align/bam.py:43]: type hinting

Comment (bsanchez-myriad) [reply Bio/Align/bam.py:43]: defining

File: Bio/Align/bam.py [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+        b"C": b"<B",
+        b"s": b"<h",
+        b"S": b"<H",
+        b"i": b"<i",
+        b"I": b"<I",
+        b"f": b"f",
+    }
+    qual_mapping: bytes = bytes(
+        (letter + 33 if 0 <= letter < 94 else 255) for letter in range(256)
+    )
```
Comment (bsanchez-myriad) [inline Bio/Align/bam.py:73]: better docs for attributes

File: Bio/Align/bam.py [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+        self._header_iterator = None
+
+    def seek(self, virtual_offset: Literal[0]) -> int:
+        if virtual_offset != 0:
+            raise NotImplementedError(
+                f"BAM reader only accepts virtual_offset = 0, but found {virtual_offset=}"
+            )
+        self._header_iterator = None
+        self._references = None
+        return super().seek(virtual_offset)
```
Comment (bsanchez-myriad) [inline Bio/Align/bam.py:89]: better docs around seek() and read()

File: Bio/Align/bam.py [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+
+    def read(self, size=-1):
+        if size != 0:
```
Comment (bsanchez-myriad) [inline Bio/Align/bam.py:92]: maybe using try and except that can catch more cases for Line 83 and line 92

File: Bio/Align/bam.py [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+    """
+
+    fmt = "BAM"
+
+    def __init__(self, source):
+        """Source should be a binary source accepted by BgzfReader"""
+        self.source = source
+        self._stream = _BAMToSamReader(source)
+        self._index = 0
+        self._read_header(self._stream)
```
Comment (bsanchez-myriad) [inline Bio/Align/bam.py:305]: Adding Unit Test

Comment (bsanchez-myriad) [reply Bio/Align/bam.py:305]: to help debug and understand the methods

Comment (bsanchez-myriad) [reply Bio/Align/bam.py:305]: Test dir using py test

File: Bio/Align/bam.py [hunk 12]
```diff
@@ -0,0 +1,305 @@
+"""Bio.Align support for the "bam" pairwise alignment format."""
```
Comment (bsanchez-myriad) [inline Bio/Align/bam.py:1]: better docs on what the file is doing to help the reviewer
